semantics/executable-spec/src/Control/Iterate/SetAlgebra.hs

```diff
@@ -0,0 +1,1250 @@
+andPQ x y p = AndPD (query x) (query y) p
+
+guardQ ::
+   (Ord k, HasQuery concrete k v) =>
+   concrete -> Fun (k -> v -> Bool) -> Query k v
+guardQ x p = GuardD (query x) p
+
+-- Don't know why this won't type check
+-- diffQ :: (Ord k, HasQuery concrete1 k v, HasQuery concrete2 k u) => concrete1 -> concrete2 -> Query k v
+-- diffQ = \ x y -> DiffD (query x) (query y)
+
+class HasQuery concrete k v where
+  query :: concrete -> Query k v
+
+instance HasQuery (Query k v) k v where
+   query xs = xs
+
+instance Ord k => HasQuery [(k,v)] k v where
+   query xs = BaseD ListR (List (sortBy (\ x y -> compare (fst x) (fst y)) xs))
+
+instance Ord k => HasQuery (Set.Set k) k () where
+   query xs = BaseD SetR (Sett xs)
+
+instance Ord k => HasQuery (Map.Map k v) k v where
+   query xs = BaseD MapR xs
+
+instance (Ord v,Ord k) => HasQuery (BiMap v k v) k v where
+   query xs = BaseD BiMapR xs
+
+instance Ord k => HasQuery (Single k v) k v where
+   query xs = BaseD SingleR xs
+
+-- =================================================
+-- Show Instances
+-- =================================================
+
+instance Show (BaseRep f k v) where
+  show MapR = "Map"
+  show SetR = "Set"
+  show ListR = "List"
+  show SingleR = "Single"
+  show BiMapR = "BiMap"
+
+instance Show (Exp t) where
+  show (Base rep x) = show rep++"?"
+  show (Dom x) = "(dom "++show x++")"
+  show (Rng x) = "(rng "++show x++")"
+  show (DRestrict x y) = "("++show x++" ◁ "++show y++")"
+  show (DExclude x y) = "("++show x++" ⋪ "++show y++")"
+  show (RRestrict x y) = "("++show x++" ▷ "++show y++")"
+  show (RExclude x y) = "("++show x++" ⋫ "++show y++")"
+  show (Elem k x) = "("++show k++" ∈ "++show x++")"
+  show (NotElem k x) = "("++show k++" ∉ "++show x++")"
+  show (Intersect x y) = "("++show x ++" ∩ "++show y++")"
+  show (Subset x y) = "("++show x++" ⊆ "++show y++")"
+  show (UnionOverrideLeft x y) = "("++show x++" ∪ "++show y++")"
+  show (UnionPlus x y) = "("++show x++" ∪+ "++show y++")"
+  show (UnionOverrideRight x y) = "("++show x++" ⨃ "++show y++")"
+  show (Singleton x y) = "(singleton _ _ )"
+  show (SetSingleton x) = "(setSingleton _ )"
+  show (KeyEqual x y) = "("++show x++" ≍ "++show y++")"
+
+ppQuery :: Query k v -> Doc
+ppQuery (BaseD rep f) = parens $ text(show rep)
+ppQuery (ProjectD f p) = parens $ text "Proj" <+> align(vsep[ppQuery f,text(show p)])
+ppQuery (AndD f g) = parens $ text "And" <+> align(vsep[ppQuery f,ppQuery g])
+ppQuery (ChainD f g p) = parens $ text "Chain" <+> align(vsep[ppQuery f,ppQuery g,text(show p)])
+ppQuery (OrD f g p) = parens $ text "Or" <+> align(vsep[ppQuery f,ppQuery g,text(show p)])
+ppQuery (GuardD f p) = parens $ text "Guard" <+> align(vsep[ppQuery f,text(show p)])
+ppQuery (DiffD f g) = parens $ text "Diff" <+> align(vsep[ppQuery f,ppQuery g])
+ppQuery (AndPD f g p) = parens $ text "AndP" <+> align(vsep[ppQuery f,ppQuery g,text(show p)])
+
+
+instance Show (Query k v) where
+   show x = show(ppQuery x)
```

Comment:
It is actually quite hard to separate this because of the dependencies of the classes. One thing we might do is make a separate module that imports this and then exports only what the users need to see. I will try to do that. I agree that most of this should be invisible to the user.

Reply:
I made a pass at dividing this up so the SetAlgebra module only contains the stuff needed to use it.